### PR TITLE
Update build to use stages (adding PHPCS to check for CS violations)

### PR DESCRIPTION
**Pull request by lcobucci**
Since travis is now using Ubuntu Trusty as default (as explained in #2837) we need to create the schema manually in order to execute the tests. I've fixed it (applying basically the same solution that was done on #2835, #2838 and #2825), however I've also changed things to use the build stages (and added PHPCS to ensure that we follow our standards).

It's important to say that I've simplified the build jobs since MySQL 5.6 is the default version on Trusty.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -19,9 +20,7 @@ before_script:
   - if [[ $TRAVIS_PHP_VERSION = '7.1' && $DB = 'sqlite' ]]; then PHPUNIT_FLAGS="--coverage-clover .clover.xml"; else PHPUNIT_FLAGS=""; fi
   - if [[ "$PHPUNIT_FLAGS" == "" ]]; then phpenv config-rm xdebug.ini || true; fi
   - if [ "$MYSQL_VERSION" == "5.7" ]; then bash ./tests/travis/install-mysql-5.7.sh; fi;
-  - if [[ "$MYSQL_VERSION" == "5.6" || "$MYSQL_VERSION" == "5.7" ]]; then mysql -e "CREATE SCHEMA doctrine_tests; GRANT ALL PRIVILEGES ON doctrine_tests.* to travis@'%'"; fi;
-  - if [[ "$MYSQL_VERSION" == "5.6" || "$MYSQL_VERSION" == "5.7" ]]; then mysql -e "CREATE SCHEMA test_create_database; GRANT ALL PRIVILEGES ON test_create_database.* to travis@'%'"; fi;
-  - if [[ "$MYSQL_VERSION" == "5.6" || "$MYSQL_VERSION" == "5.7" ]]; then mysql -e "CREATE SCHEMA test_drop_database; GRANT ALL PRIVILEGES ON test_drop_database.* to travis@'%'"; fi;
+  - if [[ "$MYSQL_VERSION" == "5.6" || "$MYSQL_VERSION" == "5.7" || "$MARIADB_VERSION" == "10.2" || "$MARIADB_VERSION" == "10.1" || "$MARIADB_VERSION" == "10.0" || "$DB" == "mysql" || "$DB" == "mysqli" ]]; then bash ./tests/travis/create-mysql-schema.sh; fi;
```
Comment (greg0ire) [inline .travis.yml]: Consider using a multiline syntax here

File: tests/travis/create-mysql-schema.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+set -ex
+
+mysql -e "CREATE SCHEMA doctrine_tests; GRANT ALL PRIVILEGES ON doctrine_tests.* to travis@'%'"
+mysql -e "CREATE SCHEMA test_create_database; GRANT ALL PRIVILEGES ON test_create_database.* to travis@'%'";
+mysql -e "CREATE SCHEMA test_drop_database; GRANT ALL PRIVILEGES ON test_drop_database.* to travis@'%'";
```
Comment (greg0ire) [inline tests/travis/create-mysql-schema.sh]: This is mostly sql. How about creating an sql file containing these 3 queries and executing in just one `mysql` call?

File: composer.json [hunk 3]
```diff
@@ -19,7 +19,9 @@
     "require-dev": {
         "phpunit/phpunit": "^6.3",
         "phpunit/phpunit-mock-objects": "!=3.2.4,!=3.2.5",
-        "symfony/console": "2.*||^3.0"
+        "symfony/console": "2.*||^3.0",
+        "doctrine/coding-standard": "^1.0",
+        "squizlabs/php_codesniffer": "^3.0"
```
Comment (greg0ire) [inline composer.json]: consider adding `sort-packages: true` to the `config` section, it will make diffs like this one less confusing.

**Comment by Ocramius** [comment]
Sorry, this is a no-go. We have 109 open PRs that would require rework if this goes through, therefore a CS change PR cannot happen.

**Comment by Majkl578** [comment]
I don't really understand why this PR was closed. Build stages still should happen and migration to Trusty is a must. While argument with open PRs is relevant, CS changes are only part of this PR and imho could be split to separate PR targetting develop. Also CS could go to _allowed failures_.

**Comment by greg0ire** [comment]
Also, there are tools can could help the 109 PRs rework, like StyleCI

**Comment by Ocramius** [comment]
The issue is with rebasing existing PRs: don't drop this burden on contributors

**Comment by Majkl578** [comment]
They still **have to** rebase to make their builds pass on Travis. And with CS changes targetting develop, it should be fine, there are only few PRs (if any) targetting develop.

**Comment by Ocramius** [comment]
@Majkl578 this patch was about `master`

**Comment by greg0ire** [comment]
@Majkl578 no we don't closing and reopening is enough in this case

**Comment by Majkl578** [comment]
@Ocramius I know, that's why proposed to move CS changes to separate PR targeting develop and keep just Travis changes, with PHPCS in *allowed_failures*.

@greg0ire It's not enough unfortunately. Each build is based on .travis.yml in current revision, not in master.

**Comment by lcobucci** [comment]
@Ocramius I'll remove the commits regarding the fixes, allow PHPCS to fail, and reopen this PR.

**Comment by lcobucci** [comment]
@Ocramius you can review this again now 😉

**Review by Ocramius — APPROVED** [review]
LGTM :ship:

**Pull request merged**